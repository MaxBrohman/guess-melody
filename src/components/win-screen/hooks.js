import {
  countResults,
  formatTime,
  resetGame,
} from '../../reducer/actions';

export const useWinStats = (answers, dispatch) => {
  const results = countResults(answers);
  const finalTime = formatTime(answers[answers.length - 1].time);
  const playAgainHandler = () => resetGame(dispatch);
  return {
    results: Object.assign({}, results, {time: finalTime}),
    playAgainHandler,
  };
};
