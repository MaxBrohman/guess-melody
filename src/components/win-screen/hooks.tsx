import {
  countResults,
  formatTime,
  resetGame,
} from '../../reducer/actions';
import {TAnswer, TDispatch, TCountedResults, TFormatTime} from '../../reducer/types';

type TUseWinStats = {
  results: TCountedResults & {time: TFormatTime};
  playAgainHandler: () => void;
}

export const useWinStats = (answers: Array<TAnswer>, dispatch: TDispatch): TUseWinStats => {
  const results = countResults(answers);
  const finalTime = formatTime(answers[answers.length - 1].time);
  const playAgainHandler = (): void => resetGame(dispatch);
  return {
    results: Object.assign({}, results, {time: finalTime}),
    playAgainHandler,
  };
};
