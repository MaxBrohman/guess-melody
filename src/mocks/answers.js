export const answers = [
  {
    isCorrect: true,
    isFast: false,
    time: 265,
  },
  {
    isCorrect: true,
    isFast: true,
    time: 255,
  },
  {
    isCorrect: false,
    isFast: true,
    time: 245,
  },
  {
    isCorrect: false,
    isFast: true,
    time: 235,
  },
  {
    isCorrect: true,
    isFast: false,
    time: 200,
  },
  {
    isCorrect: true,
    isFast: true,
    time: 190,
  },
  {
    isCorrect: true,
    isFast: true,
    time: 180,
  }
];
