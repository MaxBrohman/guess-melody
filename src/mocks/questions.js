export const questions = [
  {
    type: `genre`,
    answer: `blues`,
    id: `01`,
    options: [
      {
        src: `/01 - Nine Inch Nails - Letting Go While Holding On.mp3`,
        genre: `jazz`,
        id: `00`,
      },
      {
        src: `/02 - Nine Inch Nails - Together.mp3`,
        genre: `blues`,
        id: `01`,
      },
      {
        src: `/03 - Nine Inch Nails - Out In The Open.mp3`,
        genre: `rock`,
        id: `02`
      },
      {
        src: `/04 - Nine Inch Nails - With Faith.mp3`,
        genre: `blues`,
        id: `03`
      }
    ]
  },
  {
    type: `artist`,
    answer: `The Stooges`,
    id: `02`,
    src: `/05 - Nine Inch Nails - Apart.mp3`,
    options: [
      {
        imgSrc: `http://placehold.it/134x134`,
        artist: `Bob Marley`
      },
      {
        imgSrc: `http://placehold.it/134x134`,
        artist: `The Stooges`
      },
      {
        imgSrc: `http://placehold.it/134x134`,
        artist: `Frank Sinatra`
      }
    ]
  },
  {
    type: `genre`,
    answer: `rock`,
    id: `03`,
    options: [
      {
        src: `/01 - Nine Inch Nails - Letting Go While Holding On.mp3`,
        genre: `jazz`,
        id: `04`
      },
      {
        src: `/02 - Nine Inch Nails - Together.mp3`,
        genre: `blues`,
        id: `05`
      },
      {
        src: `/03 - Nine Inch Nails - Out In The Open.mp3`,
        genre: `rock`,
        id: `06`
      },
      {
        src: `/04 - Nine Inch Nails - With Faith.mp3`,
        genre: `rock`,
        id: `07`
      }
    ]
  }
];
