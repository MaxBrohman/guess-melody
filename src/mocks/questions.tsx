import {TQuestion} from '../reducer/types';

export const questions: Array<TQuestion> = [
  {
    type: `genre`,
    genre: `blues`,
    answers: [
      {
        src: `/01 - Nine Inch Nails - Letting Go While Holding On.mp3`,
        genre: `jazz`,
      },
      {
        src: `/02 - Nine Inch Nails - Together.mp3`,
        genre: `blues`,
      },
      {
        src: `/03 - Nine Inch Nails - Out In The Open.mp3`,
        genre: `rock`,
      },
      {
        src: `/04 - Nine Inch Nails - With Faith.mp3`,
        genre: `blues`,
      }
    ]
  },
  {
    type: `artist`,
    song: {
      artist: `The Stooges`,
      src: `/05 - Nine Inch Nails - Apart.mp3`,
    },
    answers: [
      {
        picture: `http://placehold.it/134x134`,
        artist: `Bob Marley`
      },
      {
        picture: `http://placehold.it/134x134`,
        artist: `The Stooges`
      },
      {
        picture: `http://placehold.it/134x134`,
        artist: `Frank Sinatra`
      }
    ]
  },
  {
    type: `genre`,
    genre: `rock`,
    answers: [
      {
        src: `/01 - Nine Inch Nails - Letting Go While Holding On.mp3`,
        genre: `jazz`,
      },
      {
        src: `/02 - Nine Inch Nails - Together.mp3`,
        genre: `blues`,
      },
      {
        src: `/03 - Nine Inch Nails - Out In The Open.mp3`,
        genre: `rock`,
      },
      {
        src: `/04 - Nine Inch Nails - With Faith.mp3`,
        genre: `rock`,
      }
    ]
  },
  {
    type: `artist`,
    song: {
      artist: `Frank Sinatra`,
      src: `/05 - Nine Inch Nails - Apart.mp3`,
    },
    answers: [
      {
        picture: `http://placehold.it/134x134`,
        artist: `Bob Marley`
      },
      {
        picture: `http://placehold.it/134x134`,
        artist: `The Stooges`
      },
      {
        picture: `http://placehold.it/134x134`,
        artist: `Frank Sinatra`
      }
    ]
  }
];
