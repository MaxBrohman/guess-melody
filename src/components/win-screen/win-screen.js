import React from 'react';

import {getCaption} from '../../utils/get-caption';

import {useWinStats} from './hooks';

// eslint-disable-next-line
export const WinScreen = ({answers, dispatch, mistakes, playAgainHandler}) => {
  const {results} = useWinStats(answers, dispatch);
  return (
    <section className="result">
      <div className="result__logo">
        <img src="img/melody-logo.png" alt="Угадай мелодию" width="186" height="83" />
      </div>
      <h2 className="result__title">Вы настоящий меломан!</h2>
      <p className="result__total">
        За {results.time.min} {getCaption(results.time.min, {
          one: `минуту`,
          lessThenFive: `минуты`,
          lessThenTen: `минут`,
        })} и {results.time.sec} {getCaption(results.time.sec, {
          one: `секунду`,
          lessThenFive: `секунды`,
          lessThenTen: `секунд`,
        })} вы набрали {results.points} {getCaption(results.points, {
          one: `балл`,
          lessThenFive: `балла`,
          lessThenTen: `баллов`,
        })}, дав {results.fast} {getCaption(results.fast, {
          one: `быстрый ответ`,
          lessThenFive: `быстрых ответа`,
          lessThenTen: `быстрых ответов`,
        })}, совершив {mistakes} {getCaption(mistakes, {
          one: `ошибку`,
          lessThenFive: `ошибки`,
          lessThenTen: `ошибок`,
        })}
      </p>
      <button className="replay" type="button" onClick={playAgainHandler}>Сыграть ещё раз</button>
    </section>
  );
};
