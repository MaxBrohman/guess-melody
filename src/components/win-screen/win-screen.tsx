import React from 'react';

import {getCaption} from '../../utils/get-caption';

import {useWinStats} from './hooks';
import {TAnswer, TDispatch} from '../../reducer/types';

type TWinScreenProps = {
  answers: Array<TAnswer>;
  dispatch: TDispatch;
  mistakes: number;
  playAgainHandler: () => void;
}

export const WinScreen = ({
  answers,
  dispatch,
  mistakes,
  playAgainHandler
}: TWinScreenProps): JSX.Element => {
  const {results} = useWinStats(answers, dispatch);
  return (
    <section className="result">
      <div className="result__logo">
        <img src="img/melody-logo.png" alt="Угадай мелодию" width="186" height="83" />
      </div>
      <h2 className="result__title">Вы настоящий меломан!</h2>
      <p className="result__total">
        За {results.time.min} {getCaption(results.time.min, {
          one: `минуту`,
          lessThenFive: `минуты`,
          lessThenTen: `минут`,
        })} и {results.time.sec} {getCaption(results.time.sec, {
          one: `секунду`,
          lessThenFive: `секунды`,
          lessThenTen: `секунд`,
        })} вы набрали {results.points} {getCaption(results.points, {
          one: `балл`,
          lessThenFive: `балла`,
          lessThenTen: `баллов`,
        })}, дав {results.fast} {getCaption(results.fast, {
          one: `быстрый ответ`,
          lessThenFive: `быстрых ответа`,
          lessThenTen: `быстрых ответов`,
        })}, совершив {mistakes} {getCaption(mistakes, {
          one: `ошибку`,
          lessThenFive: `ошибки`,
          lessThenTen: `ошибок`,
        })}
      </p>
      <button className="replay" type="button" onClick={playAgainHandler}>Сыграть ещё раз</button>
    </section>
  );
};
