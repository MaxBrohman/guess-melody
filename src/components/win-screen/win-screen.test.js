import React from 'react';
import renderer from 'react-test-renderer';

import {WinScreen} from './win-screen';
import {answers} from '../../mocks/answers';

it(`Renders WinScreen component correctly`, () => {
  const dispatch = jest.fn();
  const tree = renderer
                .create(
                    <WinScreen
                      answers={answers}
                      dispatch={dispatch}
                      mistakes={3}
                    />
                )
                .toJSON();
  expect(tree).toMatchSnapshot();
});
