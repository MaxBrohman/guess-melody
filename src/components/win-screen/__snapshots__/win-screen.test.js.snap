// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders WinScreen component correctly 1`] = `
<section
  className="result"
>
  <div
    className="result__logo"
  >
    <img
      alt="Угадай мелодию"
      height="83"
      src="img/melody-logo.png"
      width="186"
    />
  </div>
  <h2
    className="result__title"
  >
    Вы настоящий меломан!
  </h2>
  <p
    className="result__total"
  >
    За 
    03
     
    минуты
     и 
    00
     
    секунд
     вы набрали 
    4
     
    балла
    , дав 
    3
     
    быстрых ответа
    , совершив 
    3
     
    ошибки
  </p>
  <button
    className="replay"
    type="button"
  >
    Сыграть ещё раз
  </button>
</section>
`;
